packages/smooth_app/lib/pages/folksonomy/folksonomy_page.dart

```diff
@@ -0,0 +1,187 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+import 'package:openfoodfacts/openfoodfacts.dart';
+import 'package:provider/provider.dart';
+import 'package:smooth_app/generic_lib/bottom_sheets/smooth_bottom_sheet.dart';
+import 'package:smooth_app/generic_lib/dialogs/smooth_alert_dialog.dart';
+import 'package:smooth_app/pages/folksonomy/folksonomy_provider.dart';
+
+class FolksonomyPage extends StatelessWidget {
+  const FolksonomyPage(this.product);
+  final Product product;
+
+  @override
+  Widget build(BuildContext context) {
+    return ChangeNotifierProvider<FolksonomyProvider>(
+      create: (_) => FolksonomyProvider(product.barcode!),
+      child: FolksonomyContent(product),
+    );
+  }
+}
+
+class FolksonomyContent extends StatefulWidget {
+  const FolksonomyContent(this.product);
+  final Product product;
+
+  @override
+  FolksonomyContentState createState() => FolksonomyContentState();
+}
+
+class FolksonomyContentState extends State<FolksonomyContent> {
+  @override
+  void initState() {
+    super.initState();
+  }
+
+  void _showAddEditDialog(BuildContext context,
+      AppLocalizations appLocalizations, FolksonomyProvider provider,
+      {String? oldKey, String? oldValue}) {
+    final TextEditingController keyController =
+        TextEditingController(text: oldKey);
+    final TextEditingController valueController =
+        TextEditingController(text: oldValue);
+    const String regexString = r'^[a-z0-9_-]+(:[a-z0-9_-]+)*$';
+
+    showDialog(
+      context: context,
+      builder: (BuildContext context) {
+        return StatefulBuilder(
```

Comment:
Honestly, I'm not a fan a nesting this kind of widget.
Could you create a dedicated one?